# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- post-consensus alignment and validation with new parameter
  `--max-insertion-error`
- inserted sequences are highlighted by upper-case letters which can be
  turned off with `--no-highlight-insertions`
- batch ranges may end with a `$` indicating the end of the pileup DB
- some mechanisms for early error detection

### Changed
- cropping and splicing of insertions:
    - existing sequence is completely retained
    - moved from `process-pile-ups` to `output`
    - binary format of insertions DBs (breaking change) to gain more freedom
      in later steps
    - splice sites are chosen based on the post-consensus alignments
- ambiguities in the alignment of reads are now detected globally
- weakly anchored alignments are discarded early in the filtering pipeline
- the self- and read-alignment-based masks are now computed separately
- coverage values may now be fractional
- improved README by adhering to [Standard Readme][standard-readme]
- better (error) reporting
- temporary files have more informative names
- many minor refactorings and extensions

### Removed
- combined self- and read-alignment-based masking: old behvaiour can be copied
  by using the `--mask` parameter and supplying both masks to all commands

### Fixed
- bug that caused `check-results` to discard all alignments in certain loci


[standard-readme]: https://github.com/RichardLitt/standard-readme


## [0.0.1] - 2018-01-03
### Added
- work-around for `damapper` bug
- histograms generated by `check-results` include a column for
  `.999` sequence identity
- `check-results` optionally writes a detailed gap report

### Changed
- simplified the coverage bounds interface of `mask-repetitive-regions`: only max-values and/or the read coverage are required
- improve consensus quality by using `lasfilteralignments` to remove deteriorating local alignments
- reduced value of `--min-reads-per-pile-up` to `--min-spanning-reads` (default: 3) to better work with extremely low coverage
- reduced the default minimum anchor length to 500
- fix & simplify score function for read alignments
- suppress generation of two las files for reads alignment in
  `generate-dazzler-options`
- insertion DBs do not include information about existing contigs anymore
  which makes validation of results easier
- renamed `debug-graph` â†’ `debug-scaffold` for clarity
- added logging for discarded pile ups
- `check-results` counts complementary alignments (inversions) as errors
- remove leading/trailing gaps from all checks by `check-results`
- improved quality of documentation

## [0.0.1-rc.1] - 2018-10-15
### Added
- implemented routine to generate exact alignments from `.las` alignment chains
- better conflict handling for scaffold graph (see `--min-spanning-reads` and
  `--best-pile-up-margin`)
- resolve conflicting cropping information attached to contigs
- accept both `.db` and `.dam` files
- new pipeline command `process-pile-ups` to compute consensus in parallel
- introduced build config `testing` which produces additional commands for
  evaluation of results:
    - `translocate-gaps`
    - `find-closable-gaps`
    - `translate-coords`
    - `check-results`
- improved logging in general
- output of assembly graph using `--debug-graph`
- sorted CLI options alphabetically
- inspect Dazzler masks using `show-mask` command
- separate command `mask-repetitive-regions` for masking
- `show-*` commands have `--json` switch
- CLI switch `--usage`
- runtime improvements
- improved type definitions for better debugging experience
- many small bug fixes
